geoportailv3/static/js/draw/drawdirective.js

```diff
@@ -0,0 +1,149 @@
+goog.require('ol.interaction.Select');
+
+
+/**
+ * @param {string} appDrawTemplateUrl Url to draw template
+ * @return {angular.Directive} The Directive Definition Object.
+ * @ngInject
+ */
+app.drawDirective = function(appDrawTemplateUrl) {
+  return {
+    restrict: 'E',
+    scope: {
+      'map': '=appDrawMap',
+      'features': '=appDrawFeatures',
+      'active': '=appDrawActive'
+    },
+    controller: 'AppDrawController',
+    controllerAs: 'ctrl',
+    bindToController: true,
+    templateUrl: appDrawTemplateUrl
+  };
+};
+
+
+app.module.directive('appDraw', app.drawDirective);
+
+
+
+/**
+ * @param {!angular.Scope} $scope Scope.
+ * @param {ngeo.DecorateInteraction} ngeoDecorateInteraction Decorate
+ *     interaction service.
+ * @constructor
+ * @export
+ * @ngInject
+ */
+app.DrawController = function($scope, ngeoDecorateInteraction) {
+
+  /**
+   * @type {ol.Map}
+   *  @export
+   */
+  this.map;
+
+  /**
+   * @type {boolean}
+   *  @export
+   */
+  this.active;
+
+  /**
+   * @type {number}
+   * @private
+   */
+  this.featureSeq_ = 0;
+
+  /**
+   * @type {ol.Collection.<ol.Feature>}
+   * @export
+   */
+  this.features;
+
+  /**
+   * @type {angular.Scope}
+   * @private
+   */
+  this.scope_ = $scope;
+
+  var drawPoint = new ol.interaction.Draw({
+    features: this.features,
+    type: /** @type {ol.geom.GeometryType} */ ('Point')
+  });
+  /**
+   * @type {ol.interaction.Draw}
+   * @export
+   */
+  this.drawPoint = drawPoint;
+
+  drawPoint.setActive(false);
+  ngeoDecorateInteraction(drawPoint);
+  this.map.addInteraction(drawPoint);
+  goog.events.listen(drawPoint, ol.interaction.DrawEventType.DRAWEND,
+      angular.bind(this, this.onDrawEnd_, drawPoint));
+
+  var drawLine = new ol.interaction.Draw({
+    features: this.features,
+    type: /** @type {ol.geom.GeometryType} */ ('LineString')
+  });
+
+  /**
+   * @type {ol.interaction.Draw}
+   * @export
+   */
+  this.drawLine = drawLine;
+
+  drawLine.setActive(false);
+  ngeoDecorateInteraction(drawLine);
+  this.map.addInteraction(drawLine);
+  goog.events.listen(drawLine, ol.interaction.DrawEventType.DRAWEND,
+      angular.bind(this, this.onDrawEnd_, drawLine));
+
+
+  // Watch the "active" property, and disable the draw interactions
+  // when "active" gets set to false.
+  $scope.$watch(goog.bind(function() {
+    return this.active;
+  }, this), goog.bind(function(newVal) {
+    if (newVal === false) {
+      this.drawPoint.setActive(false);
+      this.drawLine.setActive(false);
+    }
+  }, this));
+};
+
+
+/**
+ * @param {ol.interaction.Draw} interaction
+ * @param {ol.interaction.DrawEvent} event
+ * @private
+ */
+app.DrawController.prototype.onDrawEnd_ = function(interaction, event) {
+  event.feature.set('name', 'element ' + (++this.featureSeq_));
+
+  // Deactivating asynchronosly to prevent dbl-click to zoom in
+  window.setTimeout(goog.bind(function() {
+    interaction.setActive(false);
+    this.scope_.$applyAsync();
```

Comment:
It's a better practice to do the action in a function passed to `$apply/$applyAsync`:

```js
    this.scope_.$apply(function() {
      interection.setActive(false);
    });
```

And I don't think you need to use `$applyAsync` here, you can probably just use `$apply`.